Asset manager multiple storage

Arnei opened the pull request:
> The Problem
>
>The storage of Opencast systems can easily be split between the asset manager/archive and the distributions. But splitting the archive itself is extremely hard and tricky.
>
>Unfortunately, some storage systems have limitations on how much data they can store. This means that you may have a limitation on how large your system can get based on the underlying storage system.
>
>That is why it would be incredibly helpful if Opencast could use more than one storage location/storage system for storing the asset manager data.

> Solution: Tracking Media Packages
> For this, we don’t need any migration since we would keep the same storage path structure we have in the asset manager right now:
>
> `.../archive[1..N]/<organization>/<mediapackage>/<version>/...`
>
>But to allow for dynamically adding new storage systems, we would have Opencast keep track of which media package goes to which storage, building a map like this:
>
>```
># storage(<mediapackage>) -> location
>storage("ID1") -> ".../storage/a/"
>storage("ID2") -> ".../storage/b/"
>```
>
>This means that for adding new media packages, Opencast could automatically select the storage with the lowest usage and that users could add new storage to Opencast at any time.

(lkiesow, 2022)

This PR attempts to solve the problem mentioned above by implementing the solution mentioned above.

Arnei committed:
Multiple storage directories in the asset manager

Allows for specifying multiple asset manager directories in
custom.properties.

New mediapackages are stored on the storage with the greatest usable
space.

Works with existing an asset manager directory, no migration necessary.

Arnei committed:
Configuration simplification for multiple storages

The default config option is now also used when specifying multiple
directories.

Arnei committed:
Lightly change tests for multiple asset storages

Changes the tests for the asset manager storage to also run the methods
from OsgiFileSystemAssetStore.java.

Arnei committed:
Add comments in asset manager storage

Adds some comments to function in the asset manager storage

Arnei committed:
Reimplemented StorageUsage functions for asset storage

As mentioned in the commented, the proper way to do this would be to
return the individual values for each root directory. But as far as I
can tell, that would require rewriting the StorageUsage interface, which
would then require additional changes in the workspace and working file
repository modules, which would be a lot of work for some functions that
only see use in statistics anyway.

Arnei committed:
Comment out settings that should be commented out in custom.properties

Accidentally commented in some database settings. This comments them
out again,

gregorydlogan commented:
I'm assigning myself here but if anyone else wants to be coassigned go ahead.

gregorydlogan commented on modules/asset-manager-storage-fs/src/main/java/org/opencastproject/assetmanager/storage/impl/fs/OsgiFileSystemAssetStore.java:
```diff
@@ -71,8 +92,84 @@ public class OsgiFileSystemAssetStore extends AbstractFileSystemAssetStore {
+  }
+
+  private List<String> getRootDirectories() {
+    return rootDirectories;
```
At the risk of being annoying, shouldn't this be an immutable list?

Arnei replied:
That makes sense.

gregorydlogan commented on modules/asset-manager-storage-fs/src/main/java/org/opencastproject/assetmanager/storage/impl/fs/OsgiFileSystemAssetStore.java:
```diff
@@ -103,6 +204,61 @@ public void activate(final ComponentContext cc) throws IllegalStateException, IO
+      index++;
+    }
+    for (String directory: rootDirectories) {
+      mkDirs(file(directory));
```
Hoo boy, lotsa questions :D

- What if one or more defined directories
  - Cannot be created (permissions)
  - Are *secretly* a device (cf `/dev/null`)
  - Exist but aren't writable
- Should this 
  - Does it throw an error and die? (by far the simpliest)
  - Does it continue, but not use that storage target? (ideal)
- What if one or more defined directories are *subdirectories* of one (or more!) of the others

Arnei replied:
Those ... are good questions D:

> Cannot be created (permissions)

That throws an exception

> Are secretly a device (cf /dev/null)

I don't know. And I'm not sure how to check for or handle that.

> Exist but aren't writable

That won't lead to an exception. Should/Can we check this?

> Should this [...]

Throw an error and die, no? Users probably want to know if they configured unusable storage. Or if their storage has become unusable for some reason.

> What if one or more defined directories are subdirectories of one (or more!) of the others

Oh no. Why would anyone do this. Please.  
I believe this actually should work fine for the most part, but we should probably disallow it anyway? To avoid things like someone deleting their storage and accidentally also deleting all their other storages in subdirectories

gregorydlogan replied:
> > Are secretly a device (cf /dev/null)
> 
> I don't know. And I'm not sure how to check for or handle that.

See below? I guess this isn't really likely to happen, and at a certain point PEBKAC and there's nothing we can do... 

> > Exist but aren't writable
> 
> That won't lead to an exception. Should/Can we check this?

Maybe as part of startup we try and write something to each storage? It won't guard against storages going away during runtime, but it would ensure they start out being present at least. 

> > Should this [...]
> 
> Throw an error and die, no? Users probably want to know if they configured unusable storage. Or if their storage has become unusable for some reason.

Agree, especially for something as important as this. 

> > What if one or more defined directories are subdirectories of one (or more!) of the others
> 
> Oh no. Why would anyone do this. Please.  
> I believe this actually should work fine for the most part, but we should probably disallow it anyway? To avoid things like someone deleting their storage and accidentally also deleting all their other storages in subdirectories

Because they're ~~evil~~DevOps and made an oopsie with their ansible script. I would hope that no one would do this on purpose, but I can see it happening accidentally and then not getting noticed until things get _messy_.

gregorydlogan reviewed (changes requested):
I like the idea here, and to me it reads cleanly.  I can see adopters wanting an operation to balance *current* mediapackages across their newly defined storage targets in OC 15, but that's probably outside the scope.  I have questions about the current implementation, see the review :)

Arnei committed:
Return unmodifiable list

Arnei committed:
Check if storage directories are subdirectories of each other

When configuring multiple storage directories, one might accidentally
configure a storage directory that is a subdirectory of another storage
directory. While this *could* work, it would force us to support a lot of
annoying edge cases and it is unlikely that a user would intentionally set
up his storage directories that way. So we disallow it by throwing an exception.

Arnei committed:
Check if we have write access to storage path

If we don't have write access, this throws an exception.

Arnei committed:
Appease checkstyle

gregorydlogan commented:
This, at minimum, doesn't break anything.  It also works for me (when I remember to actually mount the alternative storage directories...), so let's get this in.

Arnei commented:
Thanks for merging, but I've been informed yesterday that this (when actually used) might break the snapshot operation for older mediapackages. Currently looking into whether this is actually at fault, but seems likely so far :')

Arnei commented:
Further investigation revealed that not ALL older mediapackages were broken, just a number in the single digits. Whether that was do to this PR or some custom migration scripts remains unclear.
